src/ocrd_models/ocrd_mets.py

```diff
@@ -608,48 +607,60 @@ def get_physical_pages(self, for_fileIds : Optional[str] = None, for_pageIds : O
-                    # determine attr to look for before iterating
-                    try:
-                        attr = next(a for a in METS_PAGE_DIV_ATTRIBUTE if (
-                                    any(p in self._page_cache[a] for p in page_attr_patterns) or \
-                                    any([isinstance(p, typing.Pattern) and p.fullmatch(attr_val) \
-                                        for p in page_attr_patterns \
-                                        for attr_val in self._page_cache[a]]
-                                    )))
-                        for attr_val in self._page_cache[attr].keys():
-                            if attr_val in page_attr_patterns or \
-                                    any([isinstance(p, typing.Pattern) and p.fullmatch(attr_val) for p in page_attr_patterns]):
-                                if return_divs:
-                                    ret.append(self._page_cache[attr][attr_val])
-                                else:
-                                    ret.append(attr_val)
-                    except StopIteration:
-                        log.debug(f"No pattern matches any keys of any of the _page_caches. patterns: {page_attr_patterns}")
-                else:
-                    # determine attr during iterating
-                    attr = None
+                    page_attr_patterns.append(pageId_token)
+            if not page_attr_patterns:
+                return []
+            if self._cache_flag:
+                # determine attr to look for before iterating
+                try:
+                    for pat in page_attr_patterns:
+                        # for attr in list(METS_PAGE_DIV_ATTRIBUTE):
+                        attr : METS_PAGE_DIV_ATTRIBUTE
+                        if isinstance(pat, str):
+                            attr = next(a for a in list(METS_PAGE_DIV_ATTRIBUTE) if pat in self._page_cache[a])
+                            cache_keys = [pat]
+                        elif isinstance(pat, list):
+                            attr = next(a for a in list(METS_PAGE_DIV_ATTRIBUTE) if pat[0] in self._page_cache[a])
+                            cache_keys = [v for v in pat if v in self._page_cache[attr]]
+                        elif isinstance(pat, typing.Pattern):
+                            attr = next(a for a in list(METS_PAGE_DIV_ATTRIBUTE) for v in self._page_cache[a] if pat.fullmatch(v))
+                            cache_keys = [v for v in self._page_cache[attr] if pat.fullmatch(v)]
+                        else:
+                            raise ValueError
+                        if return_divs:
+                            ret += [self._page_cache[attr][v] for v in cache_keys]
+                        else:
+                            ret += cache_keys
+                except StopIteration:
+                    log.debug(f"No pattern matches any keys of any of the _page_caches. patterns: {page_attr_patterns}")
```

Comment:
The try-except outside the loop is plain wrong because it leads to incomplete results as you point out.

With the try-except inside the loop, a `StopIteration` means that a particular pattern does not match any pages. I think this should be a `ValueError` because it means a provided page range is completely outside what is available, or a regex is incorrect or a user-generated list of IDs, i.e. something wrong on the user side they should know about, i.e. prevent errors like https://github.com/OCR-D/core/issues/1084.

This is easy for the cached loop: move the try-except inside the loop and raise a ValueError

For the non-cached part it's a bit more involved: We need to track which patterns did match at least once and then compare whether any remain unmatched after the pattern loop. I've hacked that into 9f840671

Reply:
This change, while I still think it is right, breaks existing tests. Should I fix those and we communicate that we're being stricter now or should this just be a warning instead of a ValueError?

Reply:
> For the non-cached part it's a bit more involved: We need to track which patterns did match at least once and then compare whether any remain unmatched after the pattern loop. I've hacked that into [9f84067](https://github.com/OCR-D/core/commit/9f8406716adc33e2ba760b9f191d84f8eeb04f07)

Not necessarily: if in the uncached case the loops are inverted, then whenever a pattern remains (has not been deleted) after the page loop, this will be a ValueError.

So your https://github.com/OCR-D/core/commit/9f8406716adc33e2ba760b9f191d84f8eeb04f07 does solve it, but it could be simpler _and_ more efficient if the page/xpath loop becomes the outer loop IMO.

Reply:
Ah, so your https://github.com/OCR-D/core/pull/1063/files/9f8406716adc33e2ba760b9f191d84f8eeb04f07..264783102f36e79c14a1b53015984217ba8fb081 does invert the uncached loop. 

I'd need to look more closely, but the idea of `del`eting the patterns directly (except for the regexps of course) still seems plausible to me.

Reply:
> This change, while I still think it is right, breaks existing tests. Should I fix those and we communicate that we're being stricter now or should this just be a warning instead of a ValueError?

It does seem like an API change, yes.

To me requiring each subpattern to be matchable seems the correct thing to do. With an error message, a user can always reconsider. In contrast, with silent failure, it becomes much more difficult to ensure a certain intended result.

Reply:
> I'd need to look more closely, but the idea of `del`eting the patterns directly (except for the regexps of course) still seems plausible to me.

It does that, but defers the deletion until after the loop over the patterns, not to modify the list while iterating over it.

> Not necessarily: if in the uncached case the loops are inverted, then whenever a pattern remains (has not been deleted) after the page loop, this will be a ValueError.

That would be easier to implement but it does not work for regexen (which can match many times) nor for non-contiguous ranges, e.g. the `SBB0000F29300010000` sample has `PHYS_000{1,2,5}` and a range `PHYS_0001..PHYS_0005` would have `PHYS_000{3,4}` remaining but I think this should not be a ValueError.

Reply:
Got it. So you introduced a notion of _exhaustion_: literal matches exhaust themselves immediately, range patterns when the list has been completely consumed, and regexes never. And you remove immedately what has been matched (subpatterns and range elements alike), which is most efficient.

Further, besides exhaustion, you have _matching_: range patterns match when they have at least one matching element, and regexes when they have at least one matching page.

I did not immediately conceive of the problem that way, but it now seems compelling.

Regarding the range pattern, however, maybe there we do not want to require _any_ element to be a match, but precisely the start and the end, don't we?

Reply:
> Regarding the range pattern, however, maybe there we do not want to require any element to be a match, but precisely the start and the end, do we?

Not necessarily, I have used (in the past before METS Server and `-C -D` paritioning) the approach to make multiple copies of workspaces and then use `-g PHYS_0001..PHYS_0100` ranges in 100-page increments, not bothering to make sure that the last partition ends with the last page.

But I don't think I've ever consciously used a range with a start I expected to perhaps not match. So I could add the check for the first element to have to match.

Reply:
Yes, even although we now have the automatic chunking, users might want to split up documents that way.

I am not strongly inclined in either direction, but a check on the start pattern instead of any match would seem more inuitive.

Reply:
c6cfe03a implements a check that start of range is matched. The stubs for the same check for the last element are there but commented out because I still think it's overly strict.

Anything else still open or can we merge?

Reply:
Looks all good – thanks!!